src/examples/algorithms/AnomalyDetection.cpp

```diff
@@ -0,0 +1,218 @@
+#include <fstream> // for file output
+#include <vector>
+#include <set>
+#include <string>
+#include <cstdlib>
+
+#include "nupic/encoders/ScalarEncoder.hpp"
+#include "nupic/algorithms/SpatialPooler.hpp"
+#include "nupic/algorithms/Cells4.hpp" // TP
+#include <nupic/algorithms/TemporalMemory.hpp> // TM
+#include "nupic/algorithms/Anomaly.hpp"
+
+#include <csv.h> // external CSV parser
+#include "nupic/os/Timer.hpp"
+#include "VectorHelpers.hpp"
+
+using namespace std; // generic namespaces are included here
+using namespace nupic;
+using namespace nupic::utils;
+using namespace nupic::algorithms::spatial_pooler;
+using namespace nupic::algorithms::Cells4;
+using namespace nupic::algorithms::temporal_memory;
+using namespace nupic::algorithms::anomaly;
+
+const int DEBUG_LEVEL =1; //0=no debug (also disabled timer), ..
+
+class AnomalyDetection
+{
+
+  private:
+    ScalarEncoder encoder;
+    SpatialPooler sp;
+    Cells4 tp;
+    TemporalMemory tm;
+    Anomaly anomaly; 
+    std::vector<UInt> lastTPOutput_;
+    const string tmImpl; //"TM","TP"
+
+
+  public:
+    Real compute(Real x) {
+        auto scalar = std::vector<Real32>(encoder.getOutputWidth());
+        encoder.encodeIntoArray(x, scalar.data());
+
+        // For some reason, the ScalarEncoder outputs a Real32 vector,
+        // which is incomaptible with the SpatialPooler's input, hence a new
+        // vector is created with the required type, and the elements
+        // are converted.
+        auto uint_scalar = VectorHelpers::castVectorType<Real32, UInt>(scalar);
+        if (DEBUG_LEVEL > 1) {
+          std::cout << "Scalar encoder: ";
+          VectorHelpers::print_vector(uint_scalar);
+        }
+
+        std::vector<UInt> spOutput(sp.getNumColumns());
+        sp.compute(uint_scalar.data(), true, spOutput.data());
+        sp.stripUnlearnedColumns(spOutput.data());
+        if (DEBUG_LEVEL > 2) {
+          std::cout << "Spatial pooler: ";
+          VectorHelpers::print_vector(VectorHelpers::binaryToSparse<UInt>(spOutput), ",");
+          std::cout << std::endl;
+        }
+
+        std::vector<UInt> tpOutput;
+        if (tmImpl == "TP") {
+          std::vector<Real> rTpOutput(tp.nCells());
+          // The temporal pooler uses Real32, so the vector must be converted again
+          auto tpInput = VectorHelpers::castVectorType<UInt, Real>(spOutput);
+          tp.compute(tpInput.data(), rTpOutput.data(), true, true);
+          // And the result is converted ONCE again to UInts
+          tpOutput = VectorHelpers::castVectorType<Real32, UInt>(rTpOutput);
+        } else {
+          tm.compute(spOutput.size(), spOutput.data(), true);
+          set<UInt> both;
+          auto active = tm.getActiveCells();
+          auto pred = tm.getPredictiveCells();
+          both.insert(active.begin(), active.end());
+          both.insert(pred.begin(), pred.end());
+          tpOutput.assign(both.begin(), both.end()); //union, like TP.outputType = both
```

Comment:
@subutai I think I've seen a PR in nupic.research to "add TM output mode both like in TP" - can we add the functionality to TM? I'd use it here too.